{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import transprs as tprs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/ubuntu/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LDpred is running...\n",
      "Output file (ldcoord) already exists!  Deleting it\n",
      "\n",
      "=============================== \u001b[96mLDpred v. 1.0.11\u001b[0m ===============================\n",
      "\n",
      "Parsing summary statistics file: popEUR_SAS_merged_train.PHENO1.glm.linear\n",
      "100.00%\n",
      "SS file loaded, now sorting and storing in HDF5 file.\n",
      "Coordinating datasets (Summary statistics and LD reference genotypes).\n",
      "100.00%\n",
      "{'ldpred_action': 'coord', 'debug': False, 'gf': 'popEUR_SAS_merged_train', 'ssf': 'popEUR_SAS_merged_train.PHENO1.glm.linear', 'N': 810, 'out': 'ldcoord', 'vbim': None, 'vgf': None, 'only_hm3': False, 'ilist': None, 'skip_coordination': False, 'eff_type': 'OR', 'match_genomic_pos': False, 'maf': 0.01, 'max_freq_discrep': 0.1, 'ssf_format': 'CUSTOM', 'rs': 'ID', 'A1': 'ALT', 'A2': 'REF', 'pos': 'POS', 'info': 'INFO', 'chr': '#CHROM', 'reffreq': 'MAF', 'pval': 'P', 'eff': 'BETA', 'se': 'SE', 'ncol': 'N', 'case_freq': None, 'control_freq': None, 'case_n': None, 'control_n': None, 'z_from_se': False}\n",
      "\n",
      "========================= Summary of coordination step =========================\n",
      "Summary statistics filename:                      \n",
      "                                       popEUR_SAS_merged_train.PHENO1.glm.linear\n",
      "LD reference genotypes filename:                         popEUR_SAS_merged_train\n",
      "Coordinated data output filename:                                        ldcoord\n",
      "------------------------------ Summary statistics ------------------------------\n",
      "Num SNPs parsed from sum stats file                                        22331\n",
      "--------------------------------- Coordination ---------------------------------\n",
      "Num individuals in LD Reference data:                                        810\n",
      "SNPs in LD Reference data:                                                 22331\n",
      "Num chromosomes used:                                                          2\n",
      "SNPs common across datasets:                                               22331\n",
      "SNPs retained after filtering:                                             20832\n",
      "SNPs w ambiguous nucleotides filtered:                                      1499\n",
      "SNPs w MAF<0.010 filtered:                                                     0\n",
      "SNPs w allele freq discrepancy > 0.100 filtered:                               0\n",
      "-------------------------------- Running times ---------------------------------\n",
      "Run time for parsing summary stats:                          0 min and 10.16 sec\n",
      "Run time for coordinating datasets:                           0 min and 2.23 sec\n",
      "================================================================================\n",
      "\n",
      "\n",
      "=============================== \u001b[96mLDpred v. 1.0.11\u001b[0m ===============================\n",
      "\n",
      "Calculating LD information w. radius 500\n",
      "Storing LD information to compressed pickle file\n",
      "Applying LDpred with LD radius: 500\n",
      "20832 SNP effects were found\n",
      "Calculating LDpred-inf weights\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/ld.py:365: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  summary_dict[1.11]={'name':'Genome-wide (LDscore) estimated heritability:','value':'%0.4f'%gw_h2_ld_score_est}\n",
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/ld.py:366: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  summary_dict[1.12]={'name':'Chi-square lambda (inflation statistic).','value':'%0.4f'%chi_square_lambda}\n",
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/LDpred_inf.py:48: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  updated_betas[start_i: stop_i] = sp.dot(A_inv * n , beta_hats[start_i: stop_i])  # Adjust the beta_hats\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting LDpred gibbs with f=0.0200\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/LDpred_gibbs.py:66: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  rv_scalars[i]=sampl_var_shrink_factor* sp.sqrt((hdmp_hdmpn_i) * (1.0 / ni))\n",
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/LDpred_gibbs.py:168: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  curr_post_means[snp_i] = cd['hdmp_hdmpn'] * postp * res_beta_hat_i\n",
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/LDpred_gibbs.py:178: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  curr_betas[snp_i] = proposed_beta  # UPDATE BETA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100.00%\n",
      "\n",
      "=========================== Summary of LDpred Gibbs ============================\n",
      "Coordinated data filename                                                ldcoord\n",
      "SNP weights output file (prefix)                                         ldgibbs\n",
      "LD data filename (prefix)                                            LDF_EUR_SAS\n",
      "-------------------------------- LD information --------------------------------\n",
      "LD radius used                                                               500\n",
      "Average LD score:                                             14.089824995486646\n",
      "Genome-wide (LDscore) estimated heritability:                             0.1487\n",
      "Chi-square lambda (inflation statistic).                                  1.0815\n",
      "Running time for calculating LD information:                0 min and 12.50 secs\n",
      "SNPs in long-range LD regions                                                  0\n",
      "----------------------------- LDpred Gibbs sampler -----------------------------\n",
      "Gibbs sampler fractions used                                              [0.02]\n",
      "Number of burn-iterations used                                                10\n",
      "Number of iterations used                                                    100\n",
      "Convergence issues (for each fraction)                                    ['No']\n",
      "Running time for Gibbs sampler(s):                          2 min and 10.00 secs\n",
      "================================================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/LDpred_gibbs.py:382: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  f.write('%s    %d    %s    %s    %s    %0.4e    %0.4e\\n' % (chrom, pos, sid, nt1, nt2, raw_beta, ldpred_beta))\n",
      "/home/ubuntu/Projects/VGP-main/.conda-env/lib/python3.8/site-packages/ldpred/LDpred_gibbs.py:389: ComplexWarning: Casting complex values to real discards the imaginary part\n",
      "  f.write('%s    %d    %s    %s    %s    %0.4e    %0.4e\\n' % (chrom, pos, sid, nt1, nt2, raw_beta, ldpred_inf_beta))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output to: ldpred_score_LDpred-inf.txt\n",
      "\n",
      "=============================== \u001b[96mLDpred v. 1.0.11\u001b[0m ===============================\n",
      "\n",
      "\n",
      "Calculating LDpred-inf risk scores\n",
      "100.00%\n",
      "Writing polygenic scores to file ldpred_score_LDpred-inf.txt\n",
      "\n",
      "=============================== Scoring Summary ================================\n",
      "Validation genotype file (prefix):                                   popEUR_test\n",
      "Input weight file(s) (prefix):                                           ldgibbs\n",
      "Output scores file(s) (prefix):                                     ldpred_score\n",
      "----------------------------------- Scoring ------------------------------------\n",
      "Running time for calculating scores:                         0 min and 1.25 secs\n",
      "================================================================================\n",
      "\n",
      "--- Done in 0:02:38 ---\n"
     ]
    }
   ],
   "source": [
    "tprs.methods.ldpred(bfile_ref='popEUR_SAS_merged_train',\n",
    "       bfile_target='popEUR_test',\n",
    "       sumstat='popEUR_SAS_merged_train.PHENO1.glm.linear',\n",
    "       N=810,\n",
    "       ldr=500,\n",
    "       ldf=\"LDF_EUR_SAS\",\n",
    "       h2=0.5,\n",
    "       fraction_causal=0.02,\n",
    "       gibbs_niter=100,\n",
    "       burnin=10,\n",
    "       out=\"ldpred_score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BayesR is running...\n",
      "***********************************************\n",
      "* GCTB 2.03 beta                              *\n",
      "* Genome-wide Complex Trait Bayesian analysis *\n",
      "* Author: Jian Zeng, Luke Lloyd-Jones         *\n",
      "* MIT License                                 *\n",
      "***********************************************\n",
      "\n",
      "Analysis started: Wed Aug 11 13:28:37 2021\n",
      "\n",
      "Options:\n",
      "\n",
      "--bfile popEUR_SAS_merged_train\n",
      "--pheno pheno_train_scaled.txt\n",
      "--bayes S\n",
      "--pi 0.1\n",
      "--hsq 0.5\n",
      "--chain-length 1000\n",
      "--burn-in 50\n",
      "--out bayesR_out\n",
      "\n",
      "Reading PLINK FAM file from [popEUR_SAS_merged_train.fam].\n",
      "810 individuals to be included from [popEUR_SAS_merged_train.fam].\n",
      "Reading phenotypes from [pheno_train_scaled.txt].\n",
      "Non-missing phenotypes of trait 1 of 810 individuals are included from [pheno_train_scaled.txt].\n",
      "810 matched individuals are kept.\n",
      "Reading PLINK BIM file from [popEUR_SAS_merged_train.bim].\n",
      "22331 SNPs to be included from [popEUR_SAS_merged_train.bim].\n",
      "22331 SNPs on 2 chromosomes are included.\n",
      "Reading PLINK BED file from [popEUR_SAS_merged_train.bed] in SNP-major format ...\n",
      "Genotype data for 810 individuals and 22331 SNPs are included from [popEUR_SAS_merged_train.bed].\n",
      "\n",
      "BayesS model fitted. Algorithm: HMC.\n",
      "scale factor: 0.000283951\n",
      "MCMC launched ...\n",
      "  Chain length: 1000 iterations\n",
      "  Burn-in: 50 iterations\n",
      "\n",
      "   Iter         Pi         NnzSnp       SigmaSq       Scale          S         ResVar       GenVar         hsq          AR        Stepsize     Rounding     TimeLeft  \n",
      "    100       0.0903        2030        0.0003       0.0002       0.3573       0.8340       0.2081       0.1997       1.0000       0.0062       0.0000       0:0:18   \n",
      "    200       0.0973        2249        0.0004       0.0002       0.6991       0.7951       0.2043       0.2044       0.9900       0.0383       0.0003       0:0:20   \n",
      "    300       0.1045        2313        0.0002       0.0001       0.0659       0.8308       0.2062       0.1988       0.7967       0.0540       0.0004       0:0:16   \n",
      "    400       0.1144        2587        0.0003       0.0001       0.2529       0.7871       0.2332       0.2285       0.6900       0.0540       0.0003       0:0:15   \n",
      "    500       0.1223        2714        0.0003       0.0001       0.1922       0.6262       0.3166       0.3358       0.6440       0.0540       0.0004       0:0:13   \n",
      "    600       0.0916        1973        0.0003       0.0001       -0.4213      0.6916       0.3111       0.3102       0.6550       0.0432       0.0004       0:0:10   \n",
      "    700       0.0898        2002        0.0003       0.0001       -0.0584      0.7700       0.2082       0.2128       0.6900       0.0432       0.0003        0:0:7   \n",
      "    800       0.0876        2026        0.0003       0.0003       0.6535       0.8673       0.1220       0.1233       0.7113       0.0432       0.0003        0:0:5   \n",
      "    900       0.0709        1560        0.0003       0.0006       1.3885       0.9485       0.0542       0.0541       0.7256       0.0432       0.0003        0:0:2   \n",
      "   1000       0.0794        1770        0.0003       0.0005       1.3077       0.9198       0.0624       0.0635       0.7460       0.0432       0.0003        0:0:0   \n",
      "\n",
      "MCMC cycles completed.\n",
      "\n",
      "Posterior statistics from MCMC samples:\n",
      "\n",
      "              Mean            SD             \n",
      "        Pi    0.094626        0.014358       \n",
      "    NnzSnp    2103.842041     317.448334     \n",
      "   SigmaSq    0.000309        0.000067       \n",
      "         S    0.341652        0.453229       \n",
      "    ResVar    0.787984        0.087366       \n",
      "    GenVar    0.202353        0.084122       \n",
      "       hsq    0.203931        0.083962       \n",
      "\n",
      "Analysis finished: Wed Aug 11 13:29:04 2021\n",
      "Computational time: 0:0:27\n",
      "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to bayesR_out.log.\n",
      "Options in effect:\n",
      "  --bfile popEUR_SAS_merged_test\n",
      "  --out bayesR_out\n",
      "  --score bayesR_out.snpRes_sumstat 2 3 5\n",
      "\n",
      "7958 MB RAM detected; reserving 3979 MB for main workspace.\n",
      "22331 variants loaded from .bim file.\n",
      "204 people (0 males, 0 females, 204 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to bayesR_out.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 204 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "22331 variants and 204 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 22331 valid predictors loaded.\n",
      "--score: Results written to bayesR_out.profile .\n",
      "Output to: bayesR_out.score_BayesR\n",
      "--- Done in 0:00:27 ---\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: 1 line skipped in --score file (1 due to variant ID mismatch, 0 due to\n",
      "allele code mismatch); see bayesR_out.nopred for details.\n"
     ]
    }
   ],
   "source": [
    "tprs.methods.BayesR(\n",
    "    bfile_ref='popEUR_SAS_merged_train',\n",
    "    bfile_target='popEUR_SAS_merged_test',\n",
    "    pheno='pheno_train_scaled.txt',\n",
    "    pi=0.1,\n",
    "    h2=0.5,\n",
    "    mcmc_niter=1000,\n",
    "    burnin=50,\n",
    "    out=\"bayesR_out\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBayesR is running...\n",
      "***********************************************\n",
      "* GCTB 2.03 beta                              *\n",
      "* Genome-wide Complex Trait Bayesian analysis *\n",
      "* Author: Jian Zeng, Luke Lloyd-Jones         *\n",
      "* MIT License                                 *\n",
      "***********************************************\n",
      "\n",
      "Analysis started: Wed Aug 11 13:22:04 2021\n",
      "\n",
      "Options:\n",
      "\n",
      "--sbayes R\n",
      "--ldm popSAS_sparseldm.ldm.sparse\n",
      "--pi 0.1,0.2,0.3,0.4\n",
      "--gamma 0.04,0.05,0.2,0.1\n",
      "--gwas-summary popEUR_SAS_merged_train.PHENO1.glm.linear_sbayesr\n",
      "--exclude-mhc \n",
      "--hsq 0.5\n",
      "--chain-length 1000\n",
      "--burn-in 50\n",
      "--seed 1\n",
      "--thread 1\n",
      "--out SbayesR_out\n",
      "\n",
      "Reading SNP info from [popSAS_sparseldm.ldm.sparse.info].\n",
      "22331 SNPs to be included from [popSAS_sparseldm.ldm.sparse.info].\n",
      "0 SNPs in the MHC region (Chr6:28-34Mb) are excluded.\n",
      "Reading GWAS summary data from [popEUR_SAS_merged_train.PHENO1.glm.linear_sbayesr].\n",
      "flipped 2975 SNPs according to the minor allele in the reference and GWAS samples.\n",
      "22330 matched SNPs in the GWAS summary data (in total 22330 SNPs).\n",
      "22330 SNPs on 2 chromosomes are included.\n",
      "Reading sparse LD matrix from [popSAS_sparseldm.ldm.sparse.bin]...\n",
      "Per-SNP window size mean 100.63 sd 43.922.\n",
      "LD matrix diagonal mean 1 sd 2.81419e-06.\n",
      "Read LD matrix for 22330 SNPs (time used: 0:0:1).\n",
      "\n",
      "Data summary:\n",
      "                                             mean       sd\n",
      "            GWAS SNP Phenotypic variance    0.978    0.140\n",
      "                 GWAS SNP heterozygosity    0.397    0.104\n",
      "                    GWAS SNP sample size      810        0\n",
      "                         GWAS SNP effect    0.001    0.060\n",
      "                             GWAS SNP SE    0.057    0.010\n",
      "            MME left-hand-side diagonals  810.000    0.002\n",
      "                     MME right-hand-side    0.397   29.203\n",
      "                    LD sampling variance    1.005    0.002\n",
      "                                LD score   14.404   10.576\n",
      "\n",
      "Approximate BayesR model fitted.\n",
      "scale factor: 1.22202e-05\n",
      "Gamma: 0.04 0.05  0.2  0.1\n",
      "Fitting model assuming scaled genotypes \n",
      "MCMC launched ...\n",
      "  Chain length: 1000 iterations\n",
      "  Burn-in: 50 iterations\n",
      "\n",
      "   Iter       NumSnp1      NumSnp2      NumSnp3      NumSnp4      SigmaSq      ResVar       GenVar         hsq       Rounding     TimeLeft  \n",
      "    100        2042         4083         6166         10039       0.0000       0.9391       0.0610       0.0610       0.0000        0:0:9   \n",
      "    200        1397         4352         5841         10740       0.0000       0.9128       0.0675       0.0689       0.0001        0:0:8   \n",
      "    300        2421         2628         5931         11350       0.0000       0.9464       0.0669       0.0660       0.0001        0:0:7   \n",
      "    400        2703         2081         5887         11659       0.0000       0.8526       0.0599       0.0656       0.0001        0:0:4   \n",
      "    500        3011         2342         6229         10748       0.0000       0.9736       0.0655       0.0631       0.0001        0:0:4   \n",
      "    600        3961         1081         6072         11216       0.0000       0.8069       0.0861       0.0965       0.0001        0:0:3   \n",
      "    700        3057         1249         7061         10963       0.0000       0.9432       0.0749       0.0736       0.0001        0:0:2   \n",
      "    800        3658          912         6487         11273       0.0000       0.8313       0.0762       0.0839       0.0001        0:0:1   \n",
      "    900        2888         1049         4913         13480       0.0000       0.9275       0.0939       0.0919       0.0001        0:0:0   \n",
      "   1000        2937          940         3678         14775       0.0000       0.8405       0.0878       0.0946       0.0001        0:0:0   \n",
      "\n",
      "MCMC cycles completed.\n",
      "\n",
      "Posterior statistics from MCMC samples:\n",
      "\n",
      "              Mean            SD             \n",
      "   NumSnp1    2757.263184     711.631531     \n",
      "   NumSnp2    2163.063232     1234.616699    \n",
      "   NumSnp3    5972.884277     783.217773     \n",
      "   NumSnp4    11436.789062    1322.205688    \n",
      "   SigmaSq    0.000029        0.000004       \n",
      "    ResVar    0.896379        0.052928       \n",
      "    GenVar    0.070204        0.010642       \n",
      "       hsq    0.072838        0.011774       \n",
      "\n",
      "Analysis finished: Wed Aug 11 13:22:13 2021\n",
      "Computational time: 0:0:9\n",
      "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to SbayesR_out.log.\n",
      "Options in effect:\n",
      "  --bfile popEUR_SAS_merged_test\n",
      "  --out SbayesR_out\n",
      "  --score SbayesR_out.snpRes_sumstat 2 3 5\n",
      "\n",
      "7958 MB RAM detected; reserving 3979 MB for main workspace.\n",
      "22331 variants loaded from .bim file.\n",
      "204 people (0 males, 0 females, 204 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to SbayesR_out.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 204 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "22331 variants and 204 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 22330 valid predictors loaded.\n",
      "--score: Results written to SbayesR_out.profile .\n",
      "Output to: SbayesR_out.score_SBayesR\n",
      "--- Done in 0:00:10 ---\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: 1 line skipped in --score file (1 due to variant ID mismatch, 0 due to\n",
      "allele code mismatch); see SbayesR_out.nopred for details.\n"
     ]
    }
   ],
   "source": [
    "tprs.methods.SBayesR(\n",
    "    ldm='popSAS_sparseldm.ldm.sparse',\n",
    "    pi='0.1,0.2,0.3,0.4',\n",
    "    gamma='0.04,0.05,0.2,0.1',\n",
    "    sumstat='popEUR_SAS_merged_train.PHENO1.glm.linear',\n",
    "    bfile_target='popEUR_SAS_merged_test',\n",
    "    h2=0.5,\n",
    "    mcmc_niter=1000,\n",
    "    burnin=50,\n",
    "    af_file=\"popSAS.frq\",\n",
    "    out='SbayesR_out'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c1770e5457485421e9469f1d752ca95b38a6d919b098ff9b0b7fff618d301502"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
